sol/main.c

```diff
@@ -0,0 +1,459 @@
+
+	/* The DPDK sched library starts the bucket half full. */
+	req_queue->tb_size = rate * REQ_QUEUE_BW_RATE;
+	req_queue->tb_credits = req_queue->tb_size / 2;
+	req_queue->tb_time = 0;
+
+	req_queue->time_cpu_cycles = rte_get_tsc_cycles();
+	req_queue->time_cpu_bytes = 0;
+	req_queue->time = 0;
+
+	INIT_LIST_HEAD(&req_queue->head);
+
+	return 0;
+}
+
+static int
+cleanup_sol(struct sol_config *sol_conf)
+{
+	destroy_mailbox(&sol_conf->mb);
+	sol_conf->net = NULL;
+	if (sol_conf->gk != NULL)
+		gk_conf_put(sol_conf->gk);
+	sol_conf->gk = NULL;
+	rte_free(sol_conf);
+	return 0;
+}
+
+static int
+sol_proc(void *arg)
+{
+	struct sol_config *sol_conf = (struct sol_config *)arg;
+	unsigned int lcore = sol_conf->lcore_id;
+	uint8_t tx_port_back = sol_conf->net->back.id;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"sol: the Solicitor block is running at lcore = %u\n", lcore);
+
+	while (likely(!exiting)) {
+		enqueue_reqs(sol_conf);
+		dequeue_reqs(sol_conf, tx_port_back,
+			sol_conf->tx_queue_back);
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"sol: the Solicitor block at lcore = %u is exiting\n", lcore);
+
+	return cleanup_sol(sol_conf);
+}
+
+static int
+sol_stage1(void *arg)
+{
+	struct sol_config *sol_conf = arg;
+	int ret = get_queue_id(&sol_conf->net->back, QUEUE_TYPE_TX,
+		sol_conf->lcore_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: cannot assign a TX queue for the back interface for lcore %u\n",
+			sol_conf->lcore_id);
+		goto cleanup;
+	}
+	sol_conf->tx_queue_back = ret;
+
+	ret = init_mailbox("gk_reqs", 2 * sol_conf->pri_req_max_len,
+		sizeof(struct priority_req), sol_conf->lcore_id,
+		&sol_conf->mb);
+	if (ret < 0)
+		goto cleanup;
+
+	return 0;
+
+cleanup:
+	cleanup_sol(sol_conf);
+	return -1;
+}
+
+static int
+sol_stage2(void *arg)
+{
+	struct sol_config *sol_conf = arg;
+	int ret = req_queue_init(&sol_conf->req_queue);
+	if (ret < 0)
+		goto cleanup;
+
+	return 0;
+
+cleanup:
+	cleanup_sol(sol_conf);
+	return ret;
+}
+
+int
+run_sol(struct net_config *net_conf, struct gk_config *gk_conf,
+	struct sol_config *sol_conf)
+{
+	int ret;
+
+	if (net_conf == NULL || gk_conf == NULL || sol_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (!net_conf->back_iface_enabled) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: back interface is required\n");
+		ret = -1;
+		goto out;
+	}
+
+	ret = net_launch_at_stage1(net_conf, 0, 0, 0, 1, sol_stage1, sol_conf);
+	if (ret < 0)
+		goto out;
+
+	ret = launch_at_stage2(sol_stage2, sol_conf);
+	if (ret < 0)
+		goto stage1;
+
+	ret = launch_at_stage3("sol", sol_proc, sol_conf, sol_conf->lcore_id);
+	if (ret < 0)
+		goto stage2;
+
+	sol_conf->net = net_conf;
+	gk_conf_hold(gk_conf);
+	sol_conf->gk = gk_conf;
+
+	ret = 0;
+	goto out;
+
+stage2:
+	pop_n_at_stage2(1);
+stage1:
+	pop_n_at_stage1(1);
+out:
+	return ret;
+}
+
+/*
+ * There should be only one sol_config instance.
+ * Return an error if trying to allocate the second instance.
+ */
+struct sol_config *
+alloc_sol_conf(void)
```

Comment:
Since the Solicitor block is only needed for Gatekeeper servers, I chose to allocate it dynamically (same as the GK-GT Unit block).